File: src/renderer/db/tatoeba.ts
import pinyin from "chinese-to-pinyin";

import db from "./data";

interface ISentenceDict {
    chinese: string;
    pinyin: string;
    english: string;
}

export default function searchSentence(s: string): Promise<ISentenceDict[]> {
    return new Promise((resolve, reject) => {
        db.all(`
        SELECT
            s1."text" AS src, s2."text" AS dst
        FROM
            sentences AS s1
            INNER JOIN "links"          ON s1.id = "links".sentence_id
            INNER JOIN sentences AS s2  ON "links".translation_id = s2.id
        WHERE
            s1."text" LIKE ? AND s1.lang = ? AND s2.lang = ?`,
            [`%${s}%`, "cmn", "eng"], (err, rows) => { err ? reject(err) : resolve(rows.map((r) => {
                return {
                    chinese: r.src,
                    pinyin: pinyin(r.src),
                    english: r.dst
                } as ISentenceDict;
            })); }
        );
    });
}

File: kotlin/src/main/kotlin/hanzileveljs/Tatoeba.kt
package hanzileveljs


class Tatoeba {
    private val qBase = """
        SELECT
            s1."text" AS chinese, s2."text" AS english
        FROM
            sentence AS s1
            INNER JOIN "link"          ON s1.id = "link".sentence_id
            INNER JOIN sentence AS s2  ON "link".translation_id = s2.id
        WHERE s1.lang = :lang1 AND s2.lang = :lang2
    """.trimIndent()

    data class TatoebaEntry internal constructor(
            val chinese: String,
            val english: String
    )

    operator fun get(s: String): List<TatoebaEntry> {
        return Config.db.open().createQuery("""
            $qBase AND s1."text" LIKE :t
        """.trimIndent())
                .addParameter("t", "%$s%")
                .addParameter("lang1", "cmn")
                .addParameter("lang2", "eng")
                .executeAndFetch(TatoebaEntry::class.java)
    }

    fun random(): TatoebaEntry {
        return Config.db.open().createQuery("""
            $qBase
            ORDER BY RANDOM() LIMIT 1
        """.trimIndent())
                .addParameter("lang1", "cmn")
                .addParameter("lang2", "eng")
                .executeAndFetchFirst(TatoebaEntry::class.java)
    }
}